packages/labs/analyzer/src/lib/javascript/mixins.ts

```diff
@@ -0,0 +1,198 @@
+      `Expected mixin to have a superClass parameter.`,
+      analyzer
+    );
+    return undefined;
+  }
+  const possibleSuperClasses = fn.parameters.map((p) =>
+    ts.isIdentifier(p.name) ? p.name.text : ''
+  );
+  const functionBody = fn.body;
+  if (functionBody === undefined) {
+    addDiagnosticIfMixin(
+      fn,
+      hasMixinHint,
+      `Expected mixin to have a block function body.`,
+      analyzer
+    );
+    return undefined;
+  }
+  if (!ts.isBlock(functionBody)) {
+    addDiagnosticIfMixin(
+      fn,
+      hasMixinHint,
+      `Expected mixin to have a block function body; arrow-function class ` +
+        `expression syntax is not supported.`,
+      analyzer
+    );
+    return undefined;
+  }
+  let classDeclaration!: ts.ClassDeclaration;
+  let returnStatement!: ts.ReturnStatement;
+  functionBody.statements.forEach((s) => {
+    if (ts.isClassDeclaration(s)) {
+      classDeclaration = s;
+    }
+    if (ts.isReturnStatement(s)) {
+      returnStatement = s;
+    }
+  });
+  if (!classDeclaration) {
+    addDiagnosticIfMixin(
+      fn,
+      hasMixinHint,
+      `Expected mixin to contain a class declaration statement.`,
+      analyzer
+    );
+    return undefined;
+  }
+  if (!returnStatement) {
+    addDiagnosticIfMixin(
+      fn,
+      hasMixinHint,
+      `Expected mixin to contain a return statement returning a class.`,
+      analyzer
+    );
+    return undefined;
+  }
+  const extendsClause = classDeclaration.heritageClauses?.find(
+    (c) => c.token === ts.SyntaxKind.ExtendsKeyword
+  );
+  if (extendsClause === undefined) {
+    addDiagnosticIfMixin(
+      fn,
+      hasMixinHint,
+      `Expected mixin to contain class declaration extending a superClass argument to function.`,
+      analyzer
+    );
+    return undefined;
+  }
+  if (extendsClause.types.length !== 1) {
+    analyzer.addDiagnostic(
+      createDiagnostic({
+        typescript: analyzer.typescript,
+        node: extendsClause,
+        message:
+          'Internal error: did not expect a mixin class extends clause to have multiple types',
+        code: DiagnosticCode.UNSUPPORTED,
+        category: analyzer.typescript.DiagnosticCategory.Warning,
+      })
+    );
+    return undefined;
+  }
+  const superClassArgIdx = findSuperClassArgIndexFromHeritage(
+    possibleSuperClasses,
+    extendsClause.types[0].expression
+  );
+  if (superClassArgIdx < 0) {
+    analyzer.addDiagnostic(
+      createDiagnostic({
+        typescript: analyzer.typescript,
+        node: extendsClause,
+        message:
+          'Did not find a "superClass" argument used in the extends clause of mixin class.',
+        code: DiagnosticCode.UNSUPPORTED,
+        category: analyzer.typescript.DiagnosticCategory.Warning,
+      })
+    );
+    return undefined;
+  }
+
+  // TODO (43081j): deal with empty class declaration names properly, maybe
+  // throw an error?
+  const classDeclarationName = classDeclaration.name?.text ?? 'unknown';
```

Comment:
what do we wanna do here?

maybe if somehow the class declaration has no name, add a diagnostic and return `undefined`?

Reply:
can we dig up to the function name, and fall back to that before giving up?

```ts
export const JMixin = superClass => class extends superClass {}

// { name: 'JMixin' }
```